[RNMobile] Update image size UI control

cameronvoell opened the pull request:
## Description
<!-- Please describe what you have changed or added -->
Updates the mobile SelectControl (currently only used on mobile in the Image Block) to use a Cycle Button style UI (CyclePickerCell) instead of nested BottomSheets (PickerCell).

## How has this been tested?
See Gutenberg-Mobile PR: https://github.com/wordpress-mobile/gutenberg-mobile/pull/1574
In order to test with image urls updating correctly via api requests, this was tested in WordPress iOS and WordPress Android 
1. Open an article in the editor with an image.
2. Click the gear button to open image settings
3. Click the menu row labeled "Size" multiple times. 
4. Note the menu size value and Image within the block are updated simultaneously. 

## Screenshots
![size-changes](https://user-images.githubusercontent.com/1103838/71146730-5e891380-21f4-11ea-9eaf-6d66be5a7a34.gif)

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

- Introduces cycle button UI for updating the size of an image from the BottomSheet menu in an Image Block. 
- Removes duplicated default image size declarations and instead utilizes image sizes from the editor settings data store. 


## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->.

cameronvoell committed:
SelectControl on mobile updated to CyclePicker UI

cameronvoell committed:
Image size options now from editor settings store

cameronvoell committed:
Make CyclePickerCell variable name generic

cameronvoell commented on packages/components/src/select-control/index.native.js:
```diff
@@ -17,7 +17,7 @@ function SelectControl( {
 
 	return (
-		<PickerCell
+		<CyclePickerCell
```
Though I did not include it in this PR yet. I am proposing that we remove [PickerCell.](https://github.com/WordPress/gutenberg/blob/master/packages/components/src/mobile/bottom-sheet/picker-cell.native.js)(though Picker is still super useful!). Though PickerCell gets the BottomSheet nested menu about 80% there, it feels like a fundamental restriction of a [react-native-modal](https://github.com/react-native-community/react-native-modal) Modal nested inside of another, that we will not be able to close the containing Modal before showing the child Modal without editing the underlying library. @Tug @etoledom thoughts? 

One reason this might be worth doing now is in order to decouple PickerCell from Cross Platform component refactoring efforts, like seen here in SelectControl.

etoledom reviewed (commented):
This looks great to me and works perfect 🎉 

There's one thing I'm not sure about, maybe @pinarol and @iamthomasbishop, could give more lights:

With this PR we are changing the UX of `SelectControl`. Probably not a big deal since this is the only place we are using it so far, but for future features this could be not the best approach, if there are a lot more options to choose from.

@iamthomasbishop [already gave his 👍](https://github.com/wordpress-mobile/gutenberg-mobile/pull/1574#issuecomment-567707943) for this UX on this particular use case.

It's also true that the current `SelectControl` needs a lot of work, so this might be a good enough mid step toward a better component, but I'm unsure of the future plans. Main question is: Is it OK to replace the SelectControl UX like this or would it be better to have a different component with this UX while maintaining the current SelectControl to use in case we need it? (Still considering that PickerCell is not being removed as mentioned [here](https://github.com/WordPress/gutenberg/pull/19232#discussion_r359709858).

I personally think it's fine to merge this PR, but asking here to also give awareness of the change.

iamthomasbishop commented:
> With this PR we are changing the UX of SelectControl. Probably not a big deal since this is the only place we are using it so far, but for future features this could be not the best approach, if there are a lot more options to choose from.

@etoledom Agreed, and while this is fine for the short-term, I don't even think this is the _best_ solution for this interaction in the long-term. I don't have visibility into whether this will also change/affect other controls, but if so can we document the effects? // cc @cameronvoell

etoledom commented:
> I don't have visibility into whether this will also change/affect other controls, but if so can we document the effects?

This changes `SelectControl`, instead of presenting a new bottom-sheet with the options to select, it now behaves as shown in the gif. What I mean is that the whole control component behaves this way, not just this instance of it. But again, we don't have other instances in use anywhere else yet.

No other controls are modified at all.

cameronvoell committed:
New CycleSelectControl added

cameronvoell committed:
Merge branch 'master' into rnmobile/fix-image-size-ui-control

cameronvoell commented:
> This changes `SelectControl`, instead of presenting a new bottom-sheet with the options to select, it now behaves as shown in the gif. What I mean is that the whole control component behaves this way, not just this instance of it. But again, we don't have other instances in use anywhere else yet.

@etoledom I didn't think we were using SelectControl either, but it turns out it is in the Gallery settings menu and it was added to Video Settings two months ago here: https://github.com/WordPress/gutenberg/blob/99d38dfa3e31352dfa59dcd27b2340763835b685/packages/block-library/src/video/edit-common-settings.js#L58-L66

I added a new CycleSelectControl (https://github.com/WordPress/gutenberg/pull/19232/commits/31caac7bf44835a2c6294f30a79f6a52d967658b) to use in Image Settings until we solve the nested bottom sheet menu problem.

etoledom commented:
I've noticed [here](https://github.com/WordPress/gutenberg/pull/19800) that on Gallery Size Picker we are using the nested picker anyway.

Probably we should sync here and have the same UX in both image and gallery image sizes.

cc @iamthomasbishop @mkevins @pinarol

iamthomasbishop commented:
> Probably we should sync here and have the same UX in both image sizes and gallery image sizes

Agreed, we probably should.

Tug commented on packages/block-library/src/image/edit.native.js:
```diff
@@ -307,11 +296,18 @@ export class ImageEdit extends React.Component {
 
 		const actions = [ { label: __( 'Clear All Settings' ), onPress: this.onClearSettings } ];
 
+		const getImageSizeOptions = () => map( imageSizes, ( { label, slug } ) => {
```
I don't think it's necessary to make that a function here, just computing size options should be fine:
```
const sizeOptions = imageSizes.map( ( { label, slug } ) => ( { value: slug, label } ) );
```

cameronvoell replied:
Good call, fixed here: https://github.com/WordPress/gutenberg/pull/19232/commits/8bd9ec1071269ba8f99c78600c7942db85e7be4b

Tug commented on packages/components/src/mobile/cycle-select-control/index.native.js:
```diff
@@ -0,0 +1,35 @@
+	const id = `inspector-select-control-${ instanceId }`;
+
+	return (
+		<CyclePickerCell
```
I'm not sure I see the benefit in having both a `CycleSelectControl` and a `CyclePickerCell`, they seem mostly identical.

cameronvoell replied:
I think I'd prefer to keep it this way for now. It matches other Control / BottomSheet Cell pairs like ToggleControl / SwitchCell and I think there may be implications later on related to cross platform controls, where the control might become independent of where it lives in the UI. This issue has some interesting conversation on that topic https://github.com/WordPress/gutenberg/issues/18018

Tug commented on packages/components/src/mobile/bottom-sheet/cycle-picker-cell.native.js:
```diff
@@ -0,0 +1,30 @@
+		...cellProps
+	} = props;
+
+	const cycleOptionValue = () => {
```
A better name for this function imo would be `nextOptionValue`. Cycle does indicate that the possible options form a cycle but it's hard to understand that this is returning the next one from the currently selected one.
I think readability could be improved here as well, something as simple as this would really help I think
```
const nextOptionValue = () => {
	const selectedOptionIndex = findIndex( options, [ 'value', value ] );
	return options[ ( selectedOptionIndex + 1 ) % options.length ].value;
};
```

Tug replied:
We might want to handle the case where options is en empty array as well ;)

cameronvoell replied:
Updated constant name 9aa8aa7d5d5aad5592b40096f6ddb6e1d5c19138 . Added checks here 12ebc85ea833e3ebce644bffdf7273448c12151d  so the image block will hide image size options if sizes / default are ever empty or not matching, and a check so that the `CycleSelectControl`  `value` will just be blank if it is ever passed empty options.

Tug commented on packages/components/src/mobile/bottom-sheet/cycle-picker-cell.native.js:
```diff
@@ -0,0 +1,30 @@
+		<Cell
+			onPress={ () => onChangeValue( cycleOptionValue() ) }
+			editable={ false }
+			value={ options[ findIndex( options, [ 'value', value ] ) ].label }
```
Same here, let's create variables with explicit names that should help improve readability.

cameronvoell replied:
See comment above, this was improved in 9aa8aa7d5d5aad5592b40096f6ddb6e1d5c19138 and 12ebc85ea833e3ebce644bffdf7273448c12151d

Tug reviewed (commented):
Change looks good, I had a few minor comments about the code but otherwise I think it's ready. I like the fact that we create a new SelectControl component for that!

pinarol commented:
@mkevins this is not urgent but could you open an issue to start a discussion about whether we want to use the same circle button for Gallery or not? I believe we should be consistent among blocks but Gallery settings are x-platform already, so let's see what we can do there. We can dive into this after finishing the upload option.

etoledom commented:
Thanks @pinarol !
@cameronvoell - Let's land this PR after addressing @Tug's comments 🎉

mkevins commented:
One consideration I have for this kind of UX is for users low bandwidth / connectivity. In earlier tests with Gallery image size options, I noticed that when I change size options, I must wait for a new file to download (each size option must be cached individually). This mean that to change sizes with a cycle picker, the user may be required to download all sizes to their device, as they cycle through them. I'm not sure how much of an issue this is, or how many users will utilize this feature with low bandwidth. Perhaps not a major issue, just wanted to note that.

cameronvoell committed:
Merge branch 'master' into rnmobile/fix-image-size-ui-control

cameronvoell committed:
Replaced unneccessary getImageSizeOptions function

cameronvoell committed:
Adjusted function name and lines for readability

cameronvoell committed:
Added checks for empty size options

cameronvoell commented:
> @cameronvoell - Let's land this PR after addressing @Tug's comments 🎉

Thanks for the review @Tug! Let me know if it looks like I addressed your comments.

cameronvoell committed:
Add undefined prop protection

Tug reviewed (approved):
LGTM 👍

iamthomasbishop commented:
Sorry @mkevins, I totally missed this last week! 

> One consideration I have for this kind of UX is for users low bandwidth / connectivity. In earlier tests with Gallery image size options, I noticed that when I change size options, I must wait for a new file to download (each size option must be cached individually). This mean that to change sizes with a cycle picker, the user may be required to download all sizes to their device, as they cycle through them. I'm not sure how much of an issue this is, or how many users will utilize this feature with low bandwidth. Perhaps not a major issue, just wanted to note that. cc: @iamthomasbishop 

That's a great point, regarding offline or low-bandwidth situations. I wonder if there is any way to load only one size (large or full?) image behind the scenes and have the control only _simulate_ the change? This would bring better performance for _any_ bandwidth scenario.

etoledom commented:
> I wonder if there is any way to load only one size (large or full?) image behind the scenes and have the control only simulate the change?

Sounds like a good option! And I believe it's possible 👍 
But I'd add it as an optimisation on a separate ticket.

iamthomasbishop commented:
> Sounds like a good option! And I believe it's possible 👍
> But I'd add it as an optimisation on a separate ticket.

Sounds good! 👍

cameronvoell committed:
Merge branch 'master' into rnmobile/fix-image-size-ui-control

cameronvoell committed:
Incorporating imageSizes defaults change from master

cameronvoell commented on packages/block-library/src/image/edit.native.js:
```diff
@@ -357,9 +357,9 @@ export class ImageEdit extends React.Component {
 		];
 
-		const sizeOptions = map( imageSizes, ( { label, slug } ) => ( {
+		const sizeOptions = map( imageSizes, ( { name, slug } ) => ( {
```
Updated to accommodate this change - https://github.com/WordPress/gutenberg/pull/19800
